src/highdicom/legacy/sop.py

```diff
@@ -417,66 +545,1911 @@ def __init__(
+                        return False
+        return False
+
+    def _mark_tag_as_used(self, tg: BaseTag) -> None:
+        """Checks what group the input tag belongs to and marks it as used to
+        keep track of all used and unused tags
+        """
+        if tg in self._shared_tags:
+            self._shared_tags[tg] = True
+        elif tg in self.excluded_from_perframe_tags:
+            self.excluded_from_perframe_tags[tg] = True
+        elif tg in self._perframe_tags:
+            self._perframe_tags[tg] = True
+
+    def _copy_attrib_if_present(self, src_ds: Dataset, dest_ds: Dataset,
+                                src_kw_or_tg: str, dest_kw_or_tg: str = None,
+                                check_not_to_be_perframe: bool = True,
+                                check_not_to_be_empty: bool = False) -> None:
+        """Copies a dicom attribute value from a keyword in the source Dataset
+        to a similar or different keyword in the destination Dataset
+        Parameters
+        ----------
+        src_ds: Dataset
+            Source Dataset to copy the attribute from.
+        dest_ds: Dataset
+            Destination Dataset to copy the attribute to.
+        src_kw_or_tg: str
+            The keyword from the souce Dataset to copy its value.
+        dest_kw_or_tg: str = None
+            The keyword of the destination Dataset, the value is copied to. If
+            its value is None, then the destination keyword will be exactly the
+            source keyword.
+        check_not_to_be_perframe: bool = True
+            If this arg is true, then copy is aborted if the src attribute is
+            perframe.
+        check_not_to_be_empty: bool = False
+            If this arg is true, then copy is aborted if the source attribute is
+            empty.
+        """
+        if isinstance(src_kw_or_tg, str):
+            src_kw_or_tg = tag_for_keyword(src_kw_or_tg)
+        if dest_kw_or_tg is None:
+            dest_kw_or_tg = src_kw_or_tg
+        elif isinstance(dest_kw_or_tg, str):
+            dest_kw_or_tg = tag_for_keyword(dest_kw_or_tg)
+        if check_not_to_be_perframe:
+            if src_kw_or_tg in self._perframe_tags:
+                return
+        if src_kw_or_tg in src_ds:
+            elem = src_ds[src_kw_or_tg]
+            if check_not_to_be_empty:
+                if self._is_empty_or_empty_items(elem):
+                    return
+            new_elem = deepcopy(elem)
+            if dest_kw_or_tg == src_kw_or_tg:
+                dest_ds[dest_kw_or_tg] = new_elem
+            else:
+                new_elem1 = DataElement(dest_kw_or_tg,
+                                        dictionary_VR(dest_kw_or_tg),
+                                        new_elem.value)
+                dest_ds[dest_kw_or_tg] = new_elem1
+            # now mark the attrib as used/done to keep track of every one of it
+            self._mark_tag_as_used(src_kw_or_tg)
+
+    def _get_or_create_attribute(
```

Comment:
I wonder whether this could be drastically simplified by relying on *pydicom* using `setattr()` instead of working with "raw" `pydicom.dataelem.DataElement` instances.

Reply:
Well, we want a deep copy of an attribute from a source if available and not empty, otherwise, create a new one with a default value. In some sense, I could use setattr, but given my intent to control the Date and Time and empty attributes I would wind up writing some code more or less similar.